### DISCUSSION TITLE
ImageGlass Moon 9.0 RC 🌝

### DISCUSSION THREAD

**Discussion by d2phap** [d2phap/ImageGlass#1665] [General]
<img width="640" alt="ImageGlass 9.0 RC" src="https://github.com/ImageGlass/website-content/blob/main/releases/screenshots/v9.0-rc/9.0rc_1.jpg?raw=true">

[![Downloads](https://img.shields.io/github/downloads/d2phap/imageglass/9.0.6.1008/total?color=%232A7C91&label=downloads&style=for-the-badge)](https://imageglass.org/release/imageglass-9.0-rc-45)

Please support this project by making a donation or purchasing the Microsoft Store version (version 8):

<a href="https://www.microsoft.com/store/productId/9N33VZK3C7TH?cid=GitHubRelease&referrer=appbadge&source=GithubD2phap">
  <img height="69" src="https://user-images.githubusercontent.com/3154213/197671083-4099d80f-07fc-4af9-b80d-6357f0968347.svg" />
</a>

### 🟢 Recommended
- Windows 10/11 x64 version 22H2
- [.NET Desktop Runtime 8](https://dotnet.microsoft.com/en-us/download/dotnet/8.0)
- [WebView2 Runtime x64](https://developer.microsoft.com/en-us/microsoft-edge/webview2/#download-section) (required for App settings, Quick Setup, Check for Update, About dialog)

### 🟠 Minimum requirements
- Windows 10 x64 version 2004 (20H1)
- [.NET Desktop Runtime 6](https://dotnet.microsoft.com/en-us/download/dotnet/6.0)

# 📝 Changelog:
ImageGlass 9.0 RC is the continuation of the version 9.0 beta 5,
you should read the previous changelog here: https://github.com/d2phap/ImageGlass/releases/tag/9.0.5.726.

<details>
<summary>

### Show / hide details

</summary>

## 🟣 Breaking changes
- Changed default hotkey for:
  + Gallery (from <kbd>H</kbd> to <kbd>G</kbd>)
  + Go to (from <kbd>G</kbd> to <kbd>F</kbd>)
- Renamed setting `AllFormats` to `FileFormats`, `SinglePageFormats` to `SingleFrameFormats`
  + Remove the wildcard from the format
    - Before: `"AllFormats": "*.jpg;*.png"`
    - After: `"FileFormats": ".jpg;.png"`
    - This also applies for commands `igcmd.exe set-default-viewer <formats>` and `igcmd.exe remove-default-viewer <formats>`
- Renamed setting `ToolbarItems` to `ToolbarButtons`
- Renamed setting `InfoItems` to `ImageInfoTags`
- Renamed the `Arguments` of `Tools` setting to `Argument`
- Changed the `Arguments` type of `MouseClickActions`, `ToolbarButtons` from `string` to `string[]` (string array)
- Moved theme setting `ShowTitlebarLogo` to app setting `ShowAppLogo`

## 🟣 General changes
- Upgraded Magick.NET to [version 13.3.0](https://github.com/dlemstra/Magick.NET/releases/tag/13.3.0)
- Added support for .NET 8, ARM64 (experiment #1000) build
- Added new Color dialog that supports picking alpha value
- Added feature: File > New window
- Add setting `DisabledMenus` to disable menus of ImageGlass similar to v8
- Add setting `EnableDebug` to launch ImageGlass in debug mode
- Added `.jxl` into the list of non-animated formats by default, you can change it in the setting `SingleFrameFormats`
- Added default actions for mouse clicks similar to v8
- Added icon for toolbar button "Exit" (#1659 thanks to @xmha97)
- Fixed a vulnerability occurring with WebP files ([CVE-2023-4863](https://github.com/advisories/GHSA-j7hp-h8jx-5ppr)) (#1655)
- Fixed an issue where some formats are displayed with wrong orientation #1619
- Fixed an issue where mouse click action is not executed when the `ToggleOff` is empty
- Fixed an issue where Toolbar and Gallery background color are also updated when Backdrop style is `None`
- Fixed an issue where the Exiting Full screen mode does not restore the previous window modes (Frameless, Window fit)
- Fixed an issue where resetting the background in Settings > Appearance does not follow current theme background
- Fixed an issue where app settings UI does not load due to incorrect path of `Icons` folder (#1630)
- Fixed an issue when passing method name begins with `IG_` causes error "Parameter count mismatch"
- Fixed an issue where setting `IsIntegrated = false` cause the 2nd instance of external tool cannot be launched
- Fixed an issue where pressing <kbd>Alt+F4</kbd> does not close the window
- Fixed an issue where ImageGlass cannot open some images (#1622)

## 🟣 Quick Setup dialog
- A new Out-of-box experience wizard that replaces First-launch configurations of v8
- Added 2 setting profiles: Standard user and Professional user
- Added setting `QuickSetupVersion`, set a value greater than `9` will prevent the Quick Setup dialog from showing.

## 🟣 App Settings
- Added new setting `ShowAppIcon` to toggle the ImageGlass icon on the title bar
- Added UI for setting `ImageInfoTags`
- Added UI for setting `EditApps`:
  + Allows to assigning the editing app for multiple extensions:
    ```ts
    // in igconfig.json
    "EditApps": {
      // open .JPG and .PNG file with Photoshop
      ".jpg;.png": {
        "AppName": "Photoshop",
        "Executable": "path\\to\\photoshop.exe",
        "Argument": "<file>" // <file> is the current image path to pass to the 'Executable'
      }
    }
    ```
<img width="400" align="right" src="https://github.com/d2phap/ImageGlass/assets/3154213/1b870889-712d-4c30-927b-96cc4563ab0b" />

- Added UI for setting `ToolbarButtons`:
  + Supports drag and drop to arrange the order of the buttons
  + Supports drag and drop to add a button
  + Adds "Reset to default" button
  + Adds and edits custom toolbar button in JSON format
  + Adds action buttons to move up/down, edit, delete the toolbar button
    - Only the custom toolbar button is editable
- Added UI for File type associations setting:
  + Adds information for changing File extension icons
  + Adds information for Set/Remove default photo viewer
  + Adds UI for setting `FileFormats`
- Added a "Get help" floating button for quickly accessing documentation
- Improved UI for setting `ZoomLevels`:
  + Adds a checkbox `Use smooth zooming`
  + Adds a link button `Load default zoom levels`
- Removed "ImageGlass" on the file type when registering it as the default photo viewer (#833)

## 🟣 Viewer component
- Displays more details on the viewer when ImageGlass cannot open image file

## 🟣 Gallery component
- Fixed an issue where thumbnail cache is always reset (#1647)
- Fixed an issue where thumbnail cache size limit does not follow setting `GalleryCacheSizeInMb` (#1647)

## 🟣 ImageGlass Tools
- Changed the tool config `Arguments` to `Argument`:
  ```ts
  // in igconfig.json
  "Tools": [
    {
      "ToolId": "Tool_MyDemoApp", // a unique ID
      "ToolName": "My Demo app", // name of the tool
      "Executable": "path\\to\\the\\DemoApp.exe",
      "Argument": "<file>", // <file> is the current image path to pass to the 'Executable'
      "Hotkeys": ["X", "Ctrl+E"], // press X or Ctrl+E to toggle the tool
      "IsIntegrated": true|false // true: if the tool supports 'ImageGlass.Tools'
    }
  ]
  ```

</details>

## 🟣 Other updates for ImageGlass 9.0 RC
- Published [ImageGlass.Tools library v1.9060](https://github.com/ImageGlass/ImageGlass.Tools) to support .NET 8
- Published [ExifGlass app v1.2](https://github.com/d2phap/ExifGlass)

## 🥒 Checksums:

File | SHA-1
-- | --
ImageGlass_Moon_9.0.6.1008_RC_x64_net8.zip | `8670E00C18780AD46F31E5587702353E3F8535CD`
ImageGlass_Moon_9.0.6.1008_RC_x64_net7.zip | `2AE9FB5F94BE4A293A241FDD3C6D8E15A1504B33`
ImageGlass_Moon_9.0.6.1008_RC_x64_net6.zip | `0C3E0364C1213C77544B12ACC458D1835238D6FB`
-- | --
ImageGlass_Moon_9.0.6.1008_RC_arm64_net8.zip | `B71C540A5378994B6FEE789052E26C5C07C18B9D`

**Full Changelog**: https://github.com/d2phap/ImageGlass/compare/9.0.5.726...9.0.6.1008

<hr /><em>This discussion was created from the release <a href='https://github.com/d2phap/ImageGlass/releases/tag/9.0.6.1008'>ImageGlass Moon 9.0 RC 🌝</a>.</em>

**Comment by xmha97** [comment]
Thanks 🙏

**Comment by ghost** [comment]
This is very good update... I just uninstall v8 & Set v9.RC as default.




is this impliment? https://github.com/d2phap/ImageGlass/issues/971#issuecomment-1133560019 becuse in cant find in settings

> and also (can't find this)
> - Start with Windows (Startup)
> - Run in Background

**Reply by d2phap** [reply]
> is this impliment? https://github.com/d2phap/ImageGlass/issues/971#issuecomment-1133560019 becuse in cant find in settings

It's in app settings > General > Others. It's just a simple textbox, I am still working on documentation
![image](https://github.com/d2phap/ImageGlass/assets/3154213/2dcda6b5-a569-4d81-a27e-dd478bff5fc6)

> and also (can't find this)
> - Start with Windows (Startup)
> - Run in Background

These 2 are some of the removed features from v9.0. The reason is ImageGlass 9.0 is already fast enough not to prelaunch it.
You need to uncheck the `Start with Windows` option in v8 to remove it from Windows Startup setting.

**Reply by ghost** [reply]
@d2phap 
I think you can add all tags with [+] & [×] Sign
![IMG_20231008_113230](https://github.com/d2phap/ImageGlass/assets/120308983/19eaecf9-2f2c-4897-85db-ab39fdf59710)

**Reply by d2phap** [reply]
It will be similar to the hotkey section in Add new Tool dialog (if I have enough time)

**Comment by ghost** [comment]
Hey @d2phap check this image file if working.

![20231008-01](https://github.com/d2phap/ImageGlass/assets/120308983/afc5f03a-473c-4cad-bf5d-df293ae1f8ab)

![Untitled](https://github.com/d2phap/ImageGlass/assets/120308983/59249438-957b-4199-9061-b8ea19b94aff)

**Reply by d2phap** [reply]
I can open your image with both v9.0 RC and v8.10.
Can you give some steps for reproducing?

**Reply by ghost** [reply]
In My PC v8.10 work fine with... But v9RC not working.. I try every version (.net6,7,8..)

**Reply by ghost** [reply]
@d2phap Check this [TEST_IMAGE.zip](https://github.com/d2phap/ImageGlass/files/12973352/TEST_IMAGE.zip)
.
.
![Untitled-1](https://github.com/d2phap/ImageGlass/assets/120308983/1361a62e-88f1-4361-8d5f-f678efb85b1b)

**Reply by d2phap** [reply]
I can open it as well!
Can you share your `igconfig.json` file?

![image](https://github.com/d2phap/ImageGlass/assets/3154213/c6cb162f-31b3-4853-b28c-88679bbadc53)

**Reply by ghost** [reply]
> Can you share your igconfig.json file?

@d2phap Here the igconfig.json file [igconfig.json](https://github.com/d2phap/ImageGlass/files/13035331/igconfig.json)

**Reply by ghost** [reply]
@d2phap just try .net6 version, when click this TEST IMAGE, see this
```

Version: ImageGlass v9.0.6.1008
Release code: moon
OS: Microsoft Windows NT 10.0.19045.0 64-bit

----------------------------------------------------
Error:

The parameter is incorrect.
----------------------------------------------------

System.ComponentModel.Win32Exception (0x80070057): The parameter is incorrect.
   at DirectN.ID2D1DeviceContextExtensions.CreateBitmapFromWicBitmap[T](ID2D1DeviceContext context, IWICBitmapSource source, Nullable`1 properties)
   at DirectN.ID2D1DeviceContextExtensions.CreateBitmapFromWicBitmap(IComObject`1 context, IComObject`1 source, Nullable`1 properties)
   at D2Phap.DXHelper.ToD2D1Bitmap(IComObject`1 dc, WicBitmapSource wicSrc, Nullable`1 bitmapProps) in D:\_GITHUB\@d2phap\DXControl\Source\DXControl\DXHelper.cs:line 138
   at ImageGlass.Viewer.DXCanvas.SetImage(IgImgData imgData, UInt32 frameIndex, Boolean resetZoom, Boolean enableFading, Single initOpacity, Single opacityStep, Boolean isForPreview, Boolean autoAnimate, ImgTransform transforms) in D:\_GITHUB\@d2phap\ImageGlass\v9\Components\ImageGlass.Views\DXCanvas.cs:line 2978
   at ImageGlass.FrmMain.ShowImagePreview(String filePath, CancellationToken token) in D:\_GITHUB\@d2phap\ImageGlass\v9\ImageGlass\FrmMain.cs:line 1277
   at ImageGlass.FrmMain.HandleImageProgress_Loading(ImageLoadingEventArgs e) in D:\_GITHUB\@d2phap\ImageGlass\v9\ImageGlass\FrmMain.cs:line 1009
   at ImageGlass.FrmMain.ReportToUIThread(ProgressReporterEventArgs e) in D:\_GITHUB\@d2phap\ImageGlass\v9\ImageGlass\FrmMain.cs:line 943
   at System.Progress`1.InvokeHandlers(Object state)

```
![Screenshot (1)](https://github.com/d2phap/ImageGlass/assets/120308983/75c80525-dbb5-45c7-8e6b-9ed4263ea5aa)

**Comment by shahul01** [comment]
hope there is a favourite / heart option

**Reply by d2phap** [reply]
You can follow this https://github.com/d2phap/ImageGlass/issues/141 for future update!

**Reply by shahul01** [reply]
Thank you

**Comment by ghost** [comment]
@d2phap why is 3 versions? .net6,7,8
I think its better only one version & with some .net version check

so, if imageglass find higher version .net installed, component of appropriate version are active..
.
.
.
.

for example
.net6 installed
.net7 installed
.net8 not

imageglass automaticly use .net7(highest installed ver)
